third_party/bigframes_vendored/ibis/backends/sql/compilers/bigquery/__init__.py

```diff
@@ -0,0 +1,1106 @@
+# Contains code from https://github.com/ibis-project/ibis/blob/main/ibis/backends/sql/compilers/bigquery/__init__.py
+
+"""Module to convert from Ibis expression to SQL string."""
+
+from __future__ import annotations
+
+import decimal
+import math
+import re
+from typing import Any, TYPE_CHECKING
+
+from ibis import util
+from ibis.backends.sql.compilers.base import AggGen, NULL, SQLGlotCompiler, STAR
+from ibis.backends.sql.compilers.bigquery.udf.core import PythonToJavaScriptTranslator
+from ibis.backends.sql.datatypes import BigQueryType, BigQueryUDFType
+from ibis.backends.sql.rewrites import (
+    exclude_unsupported_window_frame_from_ops,
+    exclude_unsupported_window_frame_from_rank,
+    exclude_unsupported_window_frame_from_row_number,
+    split_select_distinct_with_order_by,
+)
+import ibis.common.exceptions as com
+from ibis.common.temporal import DateUnit, IntervalUnit, TimestampUnit, TimeUnit
+import ibis.expr.datatypes as dt
+import ibis.expr.operations as ops
+import sqlglot as sg
+from sqlglot.dialects import BigQuery
+import sqlglot.expressions as sge
+
+if TYPE_CHECKING:
+    from collections.abc import Mapping
+
+    import ibis.expr.types as ir
+
+_NAME_REGEX = re.compile(r'[^!"$()*,./;?@[\\\]^`{}~\n]+')
+
+
+_MEMTABLE_PATTERN = re.compile(
+    r"^_?ibis_(?:[A-Za-z_][A-Za-z_0-9]*)_memtable_[a-z0-9]{26}$"
+)
+
+
+def _qualify_memtable(
+    node: sge.Expression, *, dataset: str | None, project: str | None
+) -> sge.Expression:
+    """Add a BigQuery dataset and project to memtable references."""
+    if isinstance(node, sge.Table) and _MEMTABLE_PATTERN.match(node.name) is not None:
+        node.args["db"] = dataset
+        node.args["catalog"] = project
+        # make sure to quote table location
+        node = _force_quote_table(node)
+    return node
+
+
+def _remove_null_ordering_from_unsupported_window(
```

Comment:
Looks like this is duplicated in `third_party/bigframes_vendored/ibis/backends/bigquery/backend.py`. Let's create a reminder somewhere to remove whichever one is the dead code.

Reply:
Will be addressed in #827